src/Rules/Test.hs

```diff
@@ -24,49 +46,51 @@ testRules = do
-            , "-e", "ghc_unregisterised=" ++ show ghcUnregisterisedInt
-            , "-e", "ghc_with_threaded_rts=0" -- TODO: support threaded
-            , "-e", "ghc_with_dynamic_rts=0" -- TODO: support dynamic
-            , "-e", "ghc_dynamic_by_default=False" -- TODO: support dynamic
-            , "-e", "ghc_dynamic=0" -- TODO: support dynamic
-            , "-e", "ghc_with_llvm=0" -- TODO: support LLVM
-            , "-e", "in_tree_compiler=True" -- TODO: when is it equal to False?
-            , "-e", "clean_only=False" -- TODO: do we need to support True?
-            , "--configfile=testsuite/config/ghc"
-            , "--config", "compiler=" ++ show (top -/- compiler)
-            , "--config", "ghc_pkg="  ++ show (top -/- ghcPkg)
-            , "--config", "haddock="  ++ show (top -/- haddock)
-            , "--summary-file", "testsuite_summary.txt"
-            , "--threads=" ++ show threads
+        -- Prepare the timeout program.
+        need [ root -/- timeoutProgPath ]
+
+        makePath       <- builderPath $ Make ""
+        top            <- topDirectory
+        ghcPath        <- (top -/-) <$> builderPath (Ghc CompileHs Stage2)
+        unregisterised <- flag GhcUnregisterised
+
+        -- Prepare extra flags to send to the Haskell compiler.
+
+        let ifMinGhcVer ver opt = do v <- ghcCanonVersion
+                                     if ver <= v then pure opt
+                                                 else pure ""
+
+        -- Read extra argument for test from command line, like `-fvectorize`.
+        opts <- fromMaybe "" <$> (liftIO $ lookupEnv "EXTRA_HC_OPTS")
+
+        -- See: https://github.com/ghc/ghc/blob/master/testsuite/mk/test.mk#L28
+        let ghcExtraFlags = opts ++ if unregisterised
+                                       then " -optc-fno-builtin"
+                                       else ""
+
+        -- Take flags to send to the Haskell compiler from test.mk.
+        -- See: https://github.com/ghc/ghc/blob/master/testsuite/mk/test.mk#L37
+        ghcFlags <- unwords <$> sequence
+            [ pure " -dcore-lint -dcmm-lint -no-user-package-db -rtsopts"
+            , pure ghcExtraFlags
+            , ifMinGhcVer "711" "-fno-warn-missed-specialisations"
+            , ifMinGhcVer "711" "-fshow-warning-groups"
+            , ifMinGhcVer "801" "-fdiagnostics-color=never"
+            , ifMinGhcVer "801" "-fno-diagnostics-show-caret"
+            , pure "-dno-debug-output"
             ]
 
-            -- , "--config", "hp2ps="    ++ quote ("hp2ps")
-            -- , "--config", "hpc="      ++ quote ("hpc")
-            -- , "--config", "gs=$(call quote_path,$(GS))"
-            -- , "--config", "timeout_prog=$(call quote_path,$(TIMEOUT_PROGRAM))"
+        -- Set environment variables for test's Makefile.
+        liftIO $ do
+            setEnv "MAKE" makePath
+            setEnv "TEST_HC" ghcPath
+            setEnv "TEST_HC_OPTS" ghcFlags
```

Comment:
We typically set the environment as follows: 

https://github.com/snowleopard/hadrian/blob/master/src/Rules/Gmp.hs#L41-L44

Reply:
I have tried

```haskell
        -- Set environment variables for test's Makefile.
        env <- sequence
            [ builderEnvironment "MAKE" $ Make ""
            , builderEnvironment "TEST_HC" $ Ghc CompileHs Stage2
            , AddEnv "TEST_HC_OPTS" <$> runTestGhcFlags
            , AddEnv "MAKE" <$> builderPath (Make "") ]

        -- Execute the test target.
        buildWithCmdOptions env $ target (vanillaContext Stage2 compiler) RunTest [] []
```

But it doesn't work. The env variable `$MAKE` is still undefined when the python driver invoke the `$MAKE` command.

Reply:
Hmm, strange. Thanks for leaving a TODO message, I also suggest to open an issue so we don't forget about this -- the ability to set an environment for a builder is important and we should figure out the best approach for it.